### PR TITLE
feat: Accept ElementsYaml for list generator to get dynamic content from a git file (#12428)

### LINKED ISSUE #12428: ApplicationSet capability to combine git file generator with list generator to generate apps based on file content
# Motivation
I have my product environment described in git as a yaml file with an array where each item defines the location of the Helm Chart, its version and the list of value files to be applied.
I want to generate the applications based on the content of this file, one Application for each entry, in a Gitops way where the applications are kept in sync with each change of this file.

# Proposal
The list generator enhanced to accept elementsYaml, so you could use a matrix generator to inject whatever elements you want:
```
generators:
- matrix:
  generators:
  - git:
      files:
      - path: your/path.yaml
  - list:
    elementsYaml: '{{ .thingsFromFile | toYaml }}'
```

### PR DESCRIPTION
**Pull request by lsoica**
Checklist:

* [x] I've created an [enhancement proposal] and discussed it with the community https://github.com/argoproj/argo-cd/issues/12428
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 

Fixes #12428

### CODE REVIEW

**Review by crenshaw-dev — COMMENTED** [review]
I'm curious, why base 64? I think plain yaml would be better for readability.

**Comment by crenshaw-dev** [comment]
Ah, I think you're hitting this bug: https://github.com/argoproj/argo-cd/pull/12287 

Could you base your PR on mine? I'll try to get my PR merged today.

**Comment by lsoica** [comment]
@crenshaw-dev Done. Looks good.

**Comment by lsoica** [comment]
@crenshaw-dev when you get a few minutes, could you please help me with a review here ?

**Comment by crenshaw-dev** [comment]
@lsoica would you mind resolving conflicts?

**Comment by lsoica** [comment]
> @lsoica would you mind resolving conflicts?

@crenshaw-dev done!

**Review by crenshaw-dev — APPROVED** [review]
lgtm, thanks @lsoica!

File: pkg/apis/application/v1alpha1/applicationset_types.go [hunk 1]
```diff
@@ -182,8 +182,9 @@ func (g ApplicationSetTerminalGenerators) toApplicationSetNestedGenerators() []A
 
 // ListGenerator include items info
 type ListGenerator struct {
-	Elements []apiextensionsv1.JSON `json:"elements" protobuf:"bytes,1,name=elements"`
-	Template ApplicationSetTemplate `json:"template,omitempty" protobuf:"bytes,2,name=template"`
+	Elements     []apiextensionsv1.JSON `json:"elements" protobuf:"bytes,1,name=elements"`
+	ElementsYaml string                 `json:"elementsYaml,omitempty" protobuf:"bytes,2,opt,name=elementsYaml"`
```
Comment (onematchfox) [inline pkg/apis/application/v1alpha1/applicationset_types.go:186]: @lsoica @crenshaw-dev Just FYI, I think this is going to be a breaking change for existing API clients as it changes the meaning of field `2` in the generated protobuf.

File: pkg/apis/application/v1alpha1/generated.proto [hunk 2]
```diff
@@ -1010,7 +1010,9 @@ message KustomizeOptions {
 message ListGenerator {
   repeated k8s.io.apiextensions_apiserver.pkg.apis.apiextensions.v1.JSON elements = 1;
 
-  optional ApplicationSetTemplate template = 2;
+  optional string elementsYaml = 2;
```
Comment (onematchfox) [inline pkg/apis/application/v1alpha1/generated.proto:1013]: See also here. To ensure backwards compatibility with existing clients `elementsYaml` should have been appended (as field 3).

Comment (lsoica) [reply pkg/apis/application/v1alpha1/generated.proto:1013]: @onematchfox Thanks! I opened this PR https://github.com/argoproj/argo-cd/pull/13335  to address it.

**Pull request merged**